pennylane/transforms/core/transform_program.py

```diff
@@ -410,95 +481,25 @@ def prune_dynamic_transform(self, type_to_keep=1):
-                    with tf.GradientTape() as tape:
-                        gate_params = classical_function(*args, **kwargs)
-
-                    jac = tape.jacobian(gate_params, args)
-                    return jac
-
-                jac = _jacobian(*args, **kwargs)
-
-            if qnode.interface == "torch":
-                import torch  # pylint: disable=import-outside-toplevel
-
-                def _jacobian(*args, **kwargs):  # pylint: disable=unused-argument
-                    jac = torch.autograd.functional.jacobian(classical_function, args)
-                    return jac
-
-                jac = _jacobian(*args, **kwargs)
-
-            if qnode.interface in ["jax", "jax-jit"]:
-                import jax  # pylint: disable=import-outside-toplevel
-
-                argnums = 0 if argnums is None else argnums
-
-                def _jacobian(*args, **kwargs):
-                    return jax.jacobian(classical_function, argnums=argnums)(*args, **kwargs)
-
-                jac = _jacobian(*args, **kwargs)
+    # pylint: disable=too-many-arguments, too-many-positional-arguments
+    def _get_classical_jacobian(self, index: int, qnode, args, kwargs, argnums):
+        if not self[index].classical_cotransform:
+            return None
+        if qnode.interface == "jax" and "argnum" in self[index].kwargs:
+            raise qml.QuantumFunctionError(
+                "argnum does not work with the Jax interface. You should use argnums instead."
+            )
 
-            return jac
+        f = partial(_classical_preprocessing, qnode, self[:index])
+        classical_jacobian = _jac_map[qnode.interface](f, argnums, *args, **kwargs)
 
-        classical_jacobians = []
-        for index, transform in enumerate(self):
-            if transform.classical_cotransform:
-                argnum = transform._kwargs.get("argnum", None)  # pylint: disable=protected-access
-                if qnode.interface == "jax" and argnum:
-                    raise qml.QuantumFunctionError(
-                        "argnum does not work with the Jax interface. You should use argnums instead."
-                    )
-                sub_program = TransformProgram(self[0:index])
-                classical_jacobian = jacobian(
-                    classical_preprocessing, sub_program, argnums, *args, **kwargs
-                )
-                tape = qml.workflow.construct_tape(qnode, level=0)(*args, **kwargs)
-                tapes, _ = sub_program((tape,))
-                multi_tapes = len(tapes) > 1
-                if not multi_tapes:
-                    classical_jacobian = [classical_jacobian]
-                classical_jacobians.append(classical_jacobian)
-            else:
-                classical_jacobians.append(None)
-        self._classical_jacobians = classical_jacobians
-        # Reset the initial tape
-        qnode.construct(args, kwargs)
+        # autograd and tf cant handle pytrees, so need to squeeze batches
```

Comment:
Is this comment at the right place? I'm not quite seeing how it comments the code below.

Reply:
Maybe "unsqueeze" would be the right term here?

In `_classical_processing`, we squeeze out a singleton dimension if there is only one tape.  Here we have to stick it back in.

I'd love to find a better solution than this squeeze-unsqueeze combination but 🤷 .  Maybe sometime later.

Reply:
Gotcha, thanks! Yeah maybe "unsqueeze" would be more comprehensive